[JasonLCY-Temp] iP

JasonLCY-Temp opened the pull request:
# Jeeves  - Your faithful chatbot 🤵 
> "An evolution of Duke with questionable quality" - Most people, probably.
[CS2103T AY23/24 Sem1 Webpage](https://nus-cs2103-ay2324s1.github.io/website/index.html)

Jeeves (barely) helps you in the following ways:
* Keeps tracks of your tasks ~~clunkily~~
* Some other **amazing** ways that I cannot remember at this point
* The quick brown fox jumped over the lazy dog

Let's see... aside from a numbered list what other tasks do I need to implement?
1. Fenced code block (with syntax highlighting)
2. A task list
3. Inline code

Give me a hand here will you? Track my remaining tasks for me
Oh, and I suppose Jeeves can do something like this as well
- [ ] Fenced Code Block (with Syntax highlighting)
- [ ] Inline code

Hey look, a convinent `main` method just flew by
```
    public static void main(String[] args) {
        new Jeeves().run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
Remove Duke.java and Add Jeeves.java

Change the name and personality of the chatbot.

Commit:
Jeeves class, main method: Update to be responsive to user input

The program only starts and ends instantly.

There is a need for it to be able to recognize user input and respond accordingly.

Let's add the ability to read and recognize user input.
Specifically, Jeeves can recognize the 'bye' command and terminate appropriately.
All other commands will simply be echoed back.

Commit:
Add Task.java

No program logic exists for managing tasks.

A Task Object will allow for the management of tasks cleanly.

Let's add the Task class with a default constructor and getter/setters.

Commit:
Update Task class: Getter methods for task count and task id

Commit:
Update Jeeves class: Add task and list task feature

Commit:
Update Task class: Add status tracking attribute and methods

Task object attribute list needs to expanded in order to extend the
scope of Jeeves's functionality to track task status.

Commit:
Update Task class: Remove unused setter method for task description

Commit:
Update Jeeves class: Change taskList scope

To avoid unnecessary code duplication when implementing task marking
and unmarking, the task printing function has been moved to a separate
method.

The scope of taskList has also been changed to be a class-level member
in order for the program to properly function.

Commit:
Task class: Replace switch case with if else blocks

Commit:
Jeeves class: Add Marking/Unmarking function

Commit:
Update program to use the toString() String representation of a Task
object for displaying output.

Commit:
Refactor project: Packaged files, removed unused methods

In preparation of adding subclasses to Task, the current java files have
been packaged for better logical organization.

Also, unused methods have been removed for better code cleanliness.

Commit:
Add Todo, Event and Deadline classes skeleton

Commit:
Update String representation of Task objects

Commit:
Update Deadline and Event classes to be functional

Commit:
Update Jeeves class: Add support for Todo, Deadline and Event tracking

Commit:
Update testcase scripts to work with Jeeves

Commit:
Remove runtime.sh

Shell script is unused as local devices are all Windows.
Please utilize the batch file for testing.

Commit:
Update testing batch file to work with this project

Commit:
Update Jeeves class: Move scanner scope

Scanner was being recreated after every command, this is a waste of
resources.

Scanner is only created once now.

Commit:
Add test cases and expected output files

Commit:
Update Jeeves class: New message output when 0 tasks recorded

Also updated test cases and expected output files to match the new
intended behaviour

Commit:
Jeeves class: Add constants and remove magic numbers

Commit:
Update Jeeves Class: Add exception handling for "mark" command

Also updated test cases and expected outputs with new expected
behaviour

Commit:
Update Jeeves Class: Add exception handling for "unmark" command

Also updated test cases and expected outputs with new expected
behaviour

Commit:
Update Jeeves class: Add exception handling for multiple commands

Specifically
* todo
* deadline
* event

Also updated test cases and expected output with the expected behaviour

Commit:
Update Jeeves class: Change taskList data structure to ArrayList

Commit:
Update Jeeves class: Add Delete function

* Tasks can now be permanently removed from the task list.
* Exception handling added for deleted tasks to "mark" and "unmark"
commands
* Test cases and expected output updated to match new expected behaviour

Commit:
Update Jeeves class: Add datafile creation ability

Jeeves will automatically check for a saved task list and
create the file and directory path required.

Also '.gitignore' update to ignore the Task data as it does not make
sense to track it with a RCS.

Commit:
Update Jeeves to store tasks to an external data file

Updated the Task classes to be able to extract the required information
to be written to the file.
Also updated test cases with new expected behaviour following a previous
change.

Commit:
Update Jeeves to read data from external data file

Task classes edited with new constructors to handle the new required
scenario.

Commit:
Merge branch 'branch-Level-7'

Commit:
Update Jeeves to understand Dates

Deadline tasks can now officially support Dates in yyyy-MM-dd format

Commit:
Merge branch 'branch-Level-8'

Commit:
Retire deprecated redirection testing files

Commit:
Refactor Jeeves class: Extract code to classes for better abstraction

Storage class deals with loading tasks from the file and writing
to the file.

UI class deals with user interactions.

Parser class deals with interpreting the user's command and extracting
relevant tokens for later usage.

TaskList class deals with the task list and it's operations.

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Add Gradle wrappers to project

Commit:
Add JUnit tests for Parser and TaskList class

Only isNotNumber method from Parser and
getTaskListDataAsString method from TaskList has unit tests

Commit:
Update Jeeves class: Add 'find' feature

'find' allows the user to search through the descriptions of all tasks
and will display all tasks that matches the search term given

Commit:
Merge branch 'branch-Level-9'

Commit:
Update code to comply with the coding standard

Code tweaked to best fit with coding standards using manual inspection,
future improvements possible with automated check styles.

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Add Javadocs for exception classes and methods.

Refactored code to reduce unnecessary string duplication when throwing
an exception. (Error message is not required to construct the exception)

Commit:
Merge branch 'master' into branch-A-JavaDoc

Commit:
Add Javadocs across the entire code base.

Specifically:
* Jeeves class
* Parser class
* Storage class
* Ui class

Commit:
Merge branch 'branch-A-JavaDoc'

Vanessamae23 commented on src/main/java/jeeves/Jeeves.java:
```diff
@@ -0,0 +1,171 @@
+public class Jeeves {
+
+    private static final String RELATIVEPATH_DATA_DIRECTORY = "data";
+    private static final String RELATIVEPATH_DATA_FILE = "data/JeevesData.txt";
```
Should it be RELATIVE_PATH_DATA_DIRECTORY instead?

Vanessamae23 commented on src/main/java/jeeves/parser/Parser.java:
```diff
@@ -0,0 +1,221 @@
+ */
+public class Parser {
+
+    private static final int FINDCOMMAND_TODO_OFFSET = 5;
```
FIND_COMMAND_TODO_OFFSET

JasonLCY-Temp replied:
Will update in a future commit, thanks for the code review

Vanessamae23 commented on src/main/java/jeeves/parser/Parser.java:
```diff
@@ -0,0 +1,221 @@
+     * @param input The string to be checked
+     * @return Whether the string is fully comprised of integers
+     */
+    static boolean isNotNumber(String input) {
```
should it better be `isNumber()`

JasonLCY-Temp replied:
Was recommended by the IDE's static analysis to ultilize this as all of the usages would become !isNumber(), and not having the negation is apparently cleaner. Thanks for the code review.

Vanessamae23 commented on src/main/java/jeeves/parser/Parser.java:
```diff
@@ -0,0 +1,221 @@
+     *
+     * @param idString The string to be checked.
+     */
+    static void checkIfValidId(String idString) {
```
is it better `isValidId()`

JasonLCY-Temp replied:
Attempting to avoid confusing method names as boolean methods typically use the is....() format and this method is void, thanks for the review

Vanessamae23 commented on src/main/java/jeeves/parser/Parser.java:
```diff
@@ -0,0 +1,221 @@
+     *
+     * @param desc The string to be checked.
+     */
+    static void checkIfDescMissing(String desc) {
```
is this better `isDescMissing()`

JasonLCY-Temp replied:
Attempting to avoid confusing method names as boolean methods typically use the is....() format and this method is void, thanks for the review

Vanessamae23 commented on src/main/java/jeeves/parser/Parser.java:
```diff
@@ -0,0 +1,221 @@
+     *
+     * @param desc The string to be checked.
+     */
+    static void checkIfDescMissing(String desc) {
```
is this more descriptive `String description`

JasonLCY-Temp replied:
Will update in a future commit, thanks for the code review

Vanessamae23 reviewed (commented):
LGTM - Looks good to me goodjob

Commit:
Jeeves & Parser class : Rename variables as per PR review suggestions

Commit:
Update project to use checkstyle via gradle.

Fix all coding standard violations that is not 'missing javadocs'

Commit:
Remove unnecessary input redirection tests

Jeeves is transitioning to a visual GUI which makes input redirection
testing redundant.

Commit:
Add functioning GUI for Jeeves

Commit:
Fix coding style violations

Commit:
Merge branch 'branch-Level-10'

Commit:
Fix bugs related to GUI commit

There were several error messages that were not being displayed on the
GUI by Jeeves.

Not having those messages be displayed to the user greatly affects user
experience as they would not have any clue what or if anything was wrong
at all.

Let's update the code such that all error messages are properly shown
by Jeeves and no part of the code attempts to print to a command line
output.

Commit:
Add assertions to multiple classes

The program assumes that perfect input/output will be received at
several points.

This can lead to unhandled crashes if there is a developer mistake.

Let's add assertions to the program so such that cases can be
specifically detected at run time.

Commit:
Merge pull request #1 from JasonLCY-Temp/A-Assertions

Add assertions to multiple classes

Commit:
Merge branch 'master' of https://github.com/JasonLCY-Temp/ip

* 'master' of https://github.com/JasonLCY-Temp/ip:
  Add assertions to multiple classes

Commit:
Merge branch 'master' into A-CodeQuality

Commit:
Improve overall code quality

The code violates several of the recommended code quality principals
as suggested by the textbook.

Let's update the code to remove or minimize the instances of these
violations.
* Long methods
* Deep Nesting
* Magic Literals
* Multiple levels of abstraction

Commit:
Merge pull request #2 from JasonLCY-Temp/A-CodeQuality

A-CodeQuality

Commit:
Merge branch 'master' of https://github.com/JasonLCY-Temp/ip

* 'master' of https://github.com/JasonLCY-Temp/ip:
  Improve overall code quality

Commit:
Fix miscellaneous check style violations

Commit:
Add the 'Note' function

Allows the user to add notes for Jeeves to track.
This iteration does not handle storing the notes to a local file or
allow the user to track their notes.

Commit:
Update Note feature

Allows the user to now track their list of notes.
Also, Jeeves will read/write the notes from/to a local data file.

Commit:
Merge branch 'BCD-Extension'

Commit:
Add product screenshot

Commit:
Update user guide

Restyled the user guide and add explanations for the 'todo' and
'deadline feature'

Commit:
Edit User Guide

Update styling for testing purposes

Commit:
Update User Guide

For Testing purposes

Commit:
Update User Guide

Add explanations for all commands in Jeeves

Commit:
Fix code style violations

Fixed all issues found with check style other than missing JavaDocs.
